### PR TITLE
Generate metadata in manifest.

### LINKED ISSUE #1959: WithMetadata support
Code sample:

```csharp
var builder = DistributedApplication.CreateBuilder(args);
var pg = builder.AddPostgres("mypostgres")
                .WithMetadata("azd-resource-flavour", "flexibleserver")
                .WithMetadata("azd-postgres-flexible-server-sky", "P1")
                .WithMetadata("azd-resource-location", "westus3");
```

This is alread implemented in this PR: #1619 

Purpose is to support underlying communication with deployment tools.

### PR DESCRIPTION
**Pull request by mitchdenny**
Fixes: https://github.com/dotnet/aspire/issues/1959

This PR adds infrastructure to support emitting a ```"metadata":``` property for resources in the manifest. The usage is as follows:

```csharp
var builder = DistributedApplication.CreateBuilder(args);
builder.AddProject<Projects.MyApp>("myapp")
       .WithMetadata("azd-identity", "guid-for-pre-existing-identity");
```

The manifest would look like this:

```json
{
  "resources": {
    "myapp": {
      // some properties omitted for brevity.
      "type": "project.v0",
      "metadata": {
        "azd-identity": "guid-for-pre-existing-identity"
     }, 
   }
  }
}
```

Metadata properties use the JSON serializer on the value field so they can be simple types such as strings, booleans, and integers, but they can also be complex objects.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1619)

### CODE REVIEW

File: src/Aspire.Hosting/ApplicationModel/ManifestMetadataAnnotation.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /// <summary>
+    /// Name of metadata entry.
+    /// </summary>
+    public string Name { get; } = name;
+
+    /// <summary>
+    /// Value of metadata entry.
+    /// </summary>
+    public object Value { get; } = value;
```
Comment (davidfowl) [inline src/Aspire.Hosting/ApplicationModel/ManifestMetadataAnnotation.cs:21]: string?

Comment (mitchdenny) [reply src/Aspire.Hosting/ApplicationModel/ManifestMetadataAnnotation.cs:21]: On hold for now but based on latest offline chat we might still allow object serialization.

**Comment by mitchdenny** [comment]
@davidfowl merge conflict resolved. This metadata PR was ready to go in unless you can spot some issues with the current impliementation.

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 2]
```diff
@@ -168,4 +168,22 @@ public void WritePortBindingEnvironmentVariables(IResource resource)
...
+        {
+            return;
+        }
+
+        Writer.WriteStartObject("metadata");
+
+        foreach (var metadataAnnotation in metadataAnnotations)
+        {
+            Writer.WritePropertyName(metadataAnnotation.Name);
+            JsonSerializer.Serialize(Writer, metadataAnnotation.Value);
```
Comment (davidfowl) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:184]: RIP AOT 😄

File: src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs [hunk 3]
```diff
@@ -473,4 +473,24 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
...
+    /// Adds metadata to resource which is output into the manifest.
+    /// </summary>
+    /// <typeparam name="T">Type of resource.</typeparam>
+    /// <param name="builder">Resource builder.</param>
+    /// <param name="name">Name of metadata.</param>
+    /// <param name="value">Value of metadata.</param>
+    /// <returns>Resource builder.</returns>
+    public static IResourceBuilder<T> WithMetadata<T>(this IResourceBuilder<T> builder, string name, object value) where T: IResource
+    {
+        var existingAnnotation = builder.Resource.Annotations.OfType<ManifestMetadataAnnotation>().SingleOrDefault(a => a.Name == name);
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs:487]: This is going to end up killed our perf when we do testing 😄.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs (hunk 2 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:184]: We can probably inject a serializer via DI here but TBH it don't think anyone would be doing AOT on their app host, given the product we are shipping for GA there is no reason to wait for the native build when you just want to launch the app to either debug locally or produce a manifest.

**Pull request merged**